proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+calculate the total fees correctly including the PRAW fees. For purpose of
+security we assume that program never rebates the PRAW fees on the accounts.
+Wallets should be extended to use these two methods to display user the correct
+information.
+
+### Calculation of PRAW Fees and rebates.
+
+User creates a transaction by setting maximum PRAW fees in terms of
+microlamports per requested CUs (`m`) and requested CUs (`C`). Solana runtime
+will multiply `m` and `C` to calculate total maximum PRAW fees (`M`) payer is
+willing to pay, then check if user has enough balance to pay all the fees
+including PRAW fees then it will load the accounts required by the transaction.
+
+Then we iterate on accounts to calculate total PRAW fees required (`T`).
+Let (`Pi`) be PRAW fee set for ith account.
+Let number of accounts loaded be (`L`)
+
+```
+T = sum(Pi * C) for i in 0 to L.
+```
+
+If M>=T we continue to execute else the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`.
+
+Let Overpaid amount (`O = M - T`)
+
+After execution, we iterate on the rebates issued to calculate total rebated
+amount (`R`), For an account there could be multiple rebates during a
+transaction. Consider that there are `N` rebates on ith account. Each rebate
+will be represented as `Rin`.
+
+```
+R = Sum( min(Pi, max(Rin where n=0 to N)) * C )
+```
+
+We then transfer `(Pi - Ri) * C` into each writable account and `O + R` back to
+the payer.
+
+### Looking at common cases
+
+For simplicity, we have written examples in total PRAW fees instead of
+microlamports per requested CUs.
+
+#### No PRAW fees enabled
+
+* A payer does not include `SetProgramRebatableAccountWriteFees` in the
+  transaction. The transaction does not write lock any accounts with PRAW fees.
+  Then the transaction is executed without the PRAW fee feature. The payer ends
+  up paying other fees.
+
+* A payer includes `SetProgramRebatableAccountWriteFees(PRAW fees)` in the
+  transaction but none of the accounts have any PRAW fees. This case is
+  considered an overpay case. The payer balance is checked for
+  `other fees + PRAW fees`.
+  1. The payer does not have enough balance: Transaction fails with an error
+    `Insufficient Balance` and the transaction is not even scheduled for
+    execution.
+  2. The payer has enough balance then the transaction is executed and PRAW
+    fees paid are transferred back to the payer in any case.
+  
+  Note in this case
+    even if there are no PRAW fees involved the payer balance is checked
+    against PRAW fees.
+
+#### PRAW fees are enabled
+
+* Fees not paid case:
+
+  A payer does not include `SetProgramRebatableAccountWriteFees` in the
+  transaction. The transaction includes one or more accounts with PRAW fees.
+  Then the transaction is failed with an error
+  `ProgramRebatableAccountWriteFeesNotPaid`. The program is not executed at all.
+  The payer ends up paying only other fees.
```

Comment:
i think this still allow cheap and effective dos. and validator will be happy to collect the other fees from those dos activities... how about not even being scheduled for execution?

Reply:
Actually, this is the same way how other errors are handled. I don't think even it is eligible for dos. The transaction is not even executed and the user end up paying the base fees. This is the correct behavior.

Reply:
hmm, unless it's one of kinds of `InsufficientError`s, i think dos is possible like by write-locking on praw-enabled accounts and purposefully attaching cu-heavy precompile ixes.

so that account write block cu limit will be reached just with base fees with pretty low microlamports-per-cu.

also flooding many txes with slightly higher priority fee with large requested cus or write-lock full of praw accounts with alt will be enough to selectively disturb certain addresses even under stake-weighted QOS is in place...

so, i think there's many ways as long as it's included into block.

Reply:
hehe, just saw https://github.com/solana-foundation/solana-improvement-documents/pull/16#issuecomment-1611496666.

it looks like stronger protection is needed for praw.... so, i think we should better off rejecting these txes

Reply:
The transaction is not executed at all if the PRAW free requirement is not met.

Reply:
I guess you find a type should write *transaction* instead of *program*